mcs/class/System/System.Net.Sockets/Socket.cs

```diff
@@ -951,16 +951,33 @@ public bool AcceptAsync (SocketAsyncEventArgs e)
 					throw new InvalidOperationException ("AcceptSocket: The socket must not be bound or connected.");
 			}
 
-			e.curSocket = this;
-			e.Worker.Init (this, e, SocketOperation.Accept);
+			SocketAsyncResult sockares = new SocketAsyncResult (this, s => SocketAsyncEventArgsCallback (AcceptAsyncCallback, s), e, SocketOperation.Accept);
 
-			SocketAsyncResult sockares = e.Worker.result;
+			e.current_socket = this;
+			e.SetLastOperation (SocketOperationToSocketAsyncOperation (SocketOperation.Accept));
+			e.SocketError = SocketError.Success;
+			e.BytesTransferred = 0;
 
-			QueueSocketAsyncResult (readQ, e.Worker, sockares);
+			QueueIOSelectorJob (readQ, sockares.handle, new IOSelectorJob (IOOperation.Read, s => BeginAcceptCallback ((SocketAsyncResult) s), sockares));
 
 			return true;
 		}
 
+		static void AcceptAsyncCallback (SocketAsyncEventArgs e, IAsyncResult ares)
+		{
+			try {
+				e.AcceptSocket = e.current_socket.EndAccept (ares);
+			} catch (SocketException ex) {
+				e.SocketError = ex.SocketErrorCode;
+			} catch (ObjectDisposedException) {
+				e.SocketError = SocketError.OperationAborted;
+			} finally {
+				if (e.AcceptSocket == null)
+					e.AcceptSocket = new Socket (e.current_socket.AddressFamily, e.current_socket.SocketType, e.current_socket.ProtocolType, null);
```

Comment:
This code has simply been moved from SocketAsyncEventArgs in https://github.com/mono/mono/pull/2060/files#diff-b34a537bb2fafe1d526398f6dc5f54f9L288